static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+
+In an earlier version of the DAC this is what we did (copying Ben Eater's video) but the colors will dull. We also tried an op-amp but that we saw some weird visual artifacts in the output. What seemed to work best was using a [line driver](https://www.ti.com/lit/gpn/SN74AHCT1G125) to supply the resister ladder with enough current to prevent the voltage dropping too low. Now the colors are properly vivid.
+
+
+### IO Control
+To actually function as a terminal Vgaterm must receive serial signals from a connected host on one of the four terminal IO ports (USB, pin headers, RS-232 over RJ45,RS-232 over DSUB-9 Serial) and update the video buffer accordingly. It also must handle keyboard input and forward those typed characters to the connected host.
+
+{% include figure.html image="/assets/images/vgaterm-terminal-sequence.drawio.png" caption="Basic Keyboard input to terminal operation" %}
+
+For keyboard input we went over several options including falling back on [PS/2 protocol over USB](https://en.wikipedia.org/wiki/PS/2_port#Conversion_between_PS/2_and_USB) and trying to write a basic software-driven USB host with the esp32c3 GPIO pins. But we settled on just using this [USB host to UART converter](https://www.tindie.com/products/matzelectronics/ch559-usb-host-to-uart-bridge-module/) by [MatzElectronics](https://www.tindie.com/stores/matzelectronics/).
+
+Software, in a loop, waits for keyboard input or input by a connected host. When input is received on either, the CPU is interrupted. If receiving key presses, the character(s) are dequeued and passed along to the connected host. If receiving characters from the host, vgaterm updates the text buffer and cursor which updates the video memory. The updated video memory is then emitted in the next frame, showing the text to the user on screen.
+
+Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this could mean it's possible to drop characters if the CPU is overwhelmed with input. We landed on this decision because the chance of dropping characters seemed small and the display shaking and wobbling felt viscerally "broken". We can also help mitigate the liklihood of dropping characters with some UART settings ([Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control)).
```

Comment:
```suggestion
Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this could mean it's possible to drop characters if the CPU is overwhelmed with input. We landed on this decision because the chance of dropping characters seemed small and the display shaking and wobbling felt viscerally "broken". We can also help mitigate the likelihood of dropping characters with some UART settings ([Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control)).
```